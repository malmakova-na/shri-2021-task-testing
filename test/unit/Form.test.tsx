import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import events from "@testing-library/user-event";
import { Router } from "react-router";
import { Provider } from "react-redux";
import { createMemoryHistory } from 'history';
import renderer from 'react-test-renderer';

import { Application } from "../../src/client/Application";
import { CartApi } from "../../src/client/api";
import { 
    initStore, productsLoad, productsLoaded, productDetailsLoad, 
    productDetailsLoaded, checkoutComplete, checkout
} from "../../src/client/store";
import { TestExampleApi } from "./TestApi";
import { PRODUCTS, getProductById, USER, CartData, CartData_Test_1 } from './TestData';
import { CartData_Test_2 } from "../../../test/test/unit/TestData";

describe("Form тесты на: ", function() {
    const history = createMemoryHistory({
        initialEntries: ["/cart"],
        initialIndex: 0
    });
    const basename = "/hw/store";
    const api = new TestExampleApi(basename);
    const cart = new CartApi();
    cart.setState(CartData_Test_1)
    const TestExampleStore = initStore(api, cart);
    const application = (
        <Router history={history}>
            <Provider store={TestExampleStore}>
            <Application />
            </Provider>
        </Router>
    );
    it("Правильное заполнение формы", () => {
        const { getByRole} = render(application);
    
        events.type(getByRole('textbox', { name: /name/i }), USER.name);
        expect(getByRole('textbox', { name: /name/i })).toHaveValue(USER.name);

        events.type(getByRole('textbox', { name: /phone/i }), USER.phone);
        expect(getByRole('textbox', { name: /phone/i })).toHaveValue(USER.phone);

        events.type(getByRole('textbox', { name: /address/i }), USER.address);
        expect(getByRole('textbox', { name: /address/i })).toHaveValue(USER.address);

        
        //screen.logTestingPlaygroundURL();

    })
    it('Отправку пустой формы', ()=> {
        const { container, getByText } = render(application);
        const checkBtn = screen.getByRole('button', {
            name: /checkout/i
        });
        events.click(checkBtn);
        expect(getByText(/please provide your name/i)).toBeInTheDocument();
        expect(getByText(/please provide a valid phone/i)).toBeInTheDocument();
        expect(getByText(/please provide a valid address/i)).toBeInTheDocument();
        //screen.logTestingPlaygroundURL()
    })
    
    it("Отрисовку сообщения о совершении заказа", () => {
        const { container } = render(application);
        TestExampleStore.dispatch(checkoutComplete(10));
        
        expect(screen.getByText(/10/i)).toBeInTheDocument();
        
        const message = container.querySelector('div.Cart-SuccessMessage');
        expect(message).toBeInTheDocument();

        
        
        //const messageSnapshot = renderer.create(application).toJSON();
      
        //expect(messageSnapshot).toMatchSnapshot();
        //screen.logTestingPlaygroundURL()
    });
});