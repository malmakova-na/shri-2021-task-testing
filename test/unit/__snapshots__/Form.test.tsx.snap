// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form тесты на:  Отрисовку сообщения о совершении заказа 1`] = `
<div
  className="Application"
>
  <nav
    className="navbar navbar-expand-sm navbar-light bg-light"
  >
    <div
      className="container"
    >
      <a
        className="Application-Brand navbar-brand"
        href="/"
        onClick={[Function]}
      >
        Example store
      </a>
      <button
        aria-label="Toggle navigation"
        className="Application-Toggler navbar-toggler"
        onClick={[Function]}
      >
        <span
          className="navbar-toggler-icon"
        />
      </button>
      <div
        className="Application-Menu collapse navbar-collapse"
      >
        <div
          className="navbar-nav"
        >
          <a
            aria-current={null}
            className="nav-link"
            href="/catalog"
            onClick={[Function]}
          >
            Catalog
          </a>
          <a
            aria-current={null}
            className="nav-link"
            href="/delivery"
            onClick={[Function]}
          >
            Delivery
          </a>
          <a
            aria-current={null}
            className="nav-link"
            href="/contacts"
            onClick={[Function]}
          >
            Contacts
          </a>
          <a
            aria-current="page"
            className="nav-link active"
            href="/cart"
            onClick={[Function]}
            style={Object {}}
          >
            Cart
          </a>
        </div>
      </div>
    </div>
  </nav>
  <div
    className="container pt-4"
  >
    <div
      className="Cart"
    >
      <div
        className="row mb-4"
      >
        <div
          className="col"
        >
          <h1>
            Shopping cart
          </h1>
          <div
            className="row my-2"
          >
            <div
              className="col-12 col-sm-8 col-md-6"
            >
              <div
                className="Cart-SuccessMessage alert alert-success"
              >
                <h4
                  className="alert-heading"
                >
                  Well done!
                </h4>
                <p>
                  Order #
                  <strong
                    className="Cart-Number"
                  >
                    10
                  </strong>
                   has been successfully completed.
                </p>
                <hr />
                <p
                  className="mb-0"
                >
                  Please wait for confirmation of delivery.
                </p>
              </div>
            </div>
          </div>
          Cart is empty. Please select products in the 
          <a
            href="/catalog"
            onClick={[Function]}
          >
            catalog
          </a>
          .
        </div>
      </div>
    </div>
  </div>
</div>
`;
